### PR TITLE
Add the post about Hercules v4

### LINKED ISSUE #235: [PROPOSAL] Hercules v4
* Title: What's new in Hercules v4
* Author(s): Vadim
* Short description: The post should describe the cool features of src-d/hercules which appeared after [v1](https://blog.sourced.tech/post/hercules/)
* Categories: technical, science
* Deadlines: none

## Table of contents

Modern Hercules is able to:

1. Process the **whole** commit history DAG, not just the main sequence. It was very hard to implement, there were very funny and unique problems to solve, many insights into how Git works, very hardcore.
2. Load plugins. A few words about how Go plugins work.
3. Merge the results from several repositories together.
4. Use UASTs for deeper analysis, e.g. structural hotness.
5. Work with siva.

There will be cool pictures with the analysis results. Also some notes on common code snippets which were not idiomatic Go and are fixed now. + some notes on the project structure evolution, how I went from flat to tree and [defined the public interface using type aliases](https://github.com/src-d/hercules/blob/master/core.go).

## Management

This section will be filled by @campoy.

* State: (proposed | writing | written | published)
* Scheduled:
* Link to post:

## Social Media

* Wording for tweet:
* Hashtags:
* Subreddits:

*NOTE* Please write in short lines so the review is easier to do.

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Fixes #235 

Staging: https://blog-staging.srcd.run/post/hercules.v4/

Some images are missing (TODO).

Signed-off-by: Vadim Markovtsev <vadim@sourced.tech>

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@campoy @vcoisne @eiso PTAL

File: content/post/hercules.v4.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+---
+author: vadim
+date: 2018-08-29
+title: "MLonGit: Hercules v4 released"
+image: /post/hercules.v4/intro.png
+description: "src-d/hercules is an open source project started in late 2016 with the goal to speed up collecting line burndown statistics from Git repositories. It has transformed into a general purpose Git repository mining framework with several cool use cases: ownership through time, file and people embeddings, structural hotness and even comment sentiment estimation. The post presents the latest 'v4' release of Hercules and gives some insights about how Git works."
```
Comment (bzz) [inline content/post/hercules.v4.md]: `insights about how Git works` -> `insights into how Git works`?

File: content/post/hercules.v4.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+### Forks and merges
+
+The difference is the best to show rather than explain:
+
+{{% caption src="/post/hercules.v4/forks_merges.gif" %}}
+Prior to v4, Hercules did not process the full commit history and worked with it as a linear sequence.
+{{% /caption %}}
+
+This feature yields an improvement of the analysis accuracy. However, nothing
+is free unfortunately: processing all the commits can considerably slow down the analysis.
```
Comment (bzz) [inline content/post/hercules.v4.md]: a comma may be needed with the interrupted `unfortunately` - `nothing is free, unfortunately`

File: content/post/hercules.v4.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+[Yifei Feng](https://github.com/yifeif) and Hanna Revinskaya.
+How to reproduce the image:
+
+```
+hercules --burndown --burndown-people --pb https://github.com/tensorflow/tensorflow | tee tensorflow.pb | labours.py -m churn_matrix
+```
+
+#### Files, developers, functions and classes in 3D
+
+If haven't read [Your Code as a Crime Scene](https://pragprog.com/book/atcrime/your-code-as-a-crime-scene)
```
Comment (bzz) [inline content/post/hercules.v4.md]: Does this sentence miss a subject, like `I you haven't read`?

File: content/post/hercules.v4.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+We can repeat the same trick with the developers: they are couples if they commit
+to the same files. [Interactive example on Tensorflow developers.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_people.json)
+
+Finally, the trick works once again with structural units in the source code. We parse
+each file with [Babelfish](https://doc.bblf.sh) and extract anything we wish: functions,
+classes, etc. Then we apply the same coupling analysis to them: we know which are changed
+in each commit.
+
+By default the functions are extracted.
```
Comment (bzz) [inline content/post/hercules.v4.md]: missing comma `By default,` ?

File: content/post/hercules.v4.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+```
+hercules --shotness --pb https://github.com/tensorflow/tensorflow | labours.py -m shotness -o tensorflow
+```
+
+#### Comment sentiment
+
+This is a proof-of-concept for running a Tensorflow model over the source code.
+We take [BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) and apply it to comments.
+That model is general purpose and the result is often weird, but it works.
```
Comment (bzz) [inline content/post/hercules.v4.md]: `That model is general purpose and` -> `That model is a general purpose one and`?

File: content/post/hercules.v4.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+a better list of changes with some of the "delete + create" pairs replaced with
+single "edit under new name" elements. Yep, Git packfiles do not store this kind
+of information (intentionally) and `git` detects renames using heuristics every
+time you execute it.
+
+Hercules builds the DAG of pipeline items automatically based on their dependencies
+and provided intermediate results. Some items try to improve on the others -
+`RenameAnalysis` for example, and Hercules understands that and correctly resolves
+the structure: all the rest of the items which depend on that type of analysis are
+pointed from the "improved" ones, and the "improved" ones in turn depend on the
```
Comment (bzz) [inline content/post/hercules.v4.md]: It appears that commas around interrupter `in turn` may be needed -> `, in turn, `

File: content/post/hercules.v4.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+It is also possible to load several serialized results from disk and combine
+them together - for example, to obtain the grand view across the company's repositories.
+
+This versatile architecture allows Hercules to efficiently analyze small and big
+Git repositories while minimizing the complexity and boilerplate code to add
+new analysis types.
+
+# v4 challenges
+
+There were three most difficult technical challenges which had to solved in v3 and v4:
```
Comment (bzz) [inline content/post/hercules.v4.md]: `had to solved` -> `had to be solved`?

File: content/post/hercules.v4.md [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+annotate each line of the each merged file, and copy the only real last edit date
+over those marked as outliers. In places of same edits but conflicting dates we
+choose the oldest date. In places of merge conflicts we automatically get the date
+of the merge commit without special tricks. We should not compare the files
+which are not involved in the merge, so we keep track of their hashes, we maintain
+them. I also had to rewrite the statistics storage logic which used to be too
+"sequential". Overall, fulfilling (2) was a technical nightmare.
+
+Surprisingly, however, most of the efforts went to (3). The first problem was
+related to minimizing the amount of forks and merges, in particular removing
```
Comment (bzz) [inline content/post/hercules.v4.md]: as forks are countable, it may be better use a different quantifier  `minimizing the amount of forks` -> `minimizing the number of forks`

File: content/post/hercules.v4.md [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+the no-op fast-forward-like back edges of the DAG.
+
+{{% caption src="/post/hercules.v4/ff.gif" %}}
+Example of the removal of a fast-forward DAG edge.
+{{% /caption %}}
+
+We [topologically sort](https://en.wikipedia.org/wiki/Topological_sorting)
+the nodes of our DAG and traverse it starting from the root.
+Each time we suspect that an edge can be removed we traverse backwards from the children
```
Comment (bzz) [inline content/post/hercules.v4.md]: I may be wrong, but American English seems to use a different spelling `traverse backwards` -> `traverse backward`. Because reasons...

File: content/post/hercules.v4.md [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+{{% caption src="/post/hercules.v4/gitk.png" %}}
+Example of multiple DAG roots: [5569bf9](https://github.com/git/git/commit/5569bf9bbedd63a00780fc5c110e0cfab3aa97b9).
+{{% /caption %}}
+
+The third problem was related to how Hercules works: it is able to analyze any
+set of commits, not just the whole master. There is not guarantee that the commits
+are connected at all, so we have to run the [connected components analysis](https://en.wikipedia.org/wiki/Connected-component_labeling)
+first to throw away everything but the main component (which has the biggest number of nodes).
+
+Thr fourth problem was to schedule the garbage collection - remove the
```
Comment (bzz) [inline content/post/hercules.v4.md]: `Thr` -> `The` ?

File: content/post/hercules.v4.md [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+The second problem was that the Git history may have more than one root commit so
+we need to find the main root and take into account the side effects.
+Don't believe me? Here is how [gitk](https://git-scm.com/docs/gitk) was merged inside git/git:
+
+{{% caption src="/post/hercules.v4/gitk.png" %}}
+Example of multiple DAG roots: [5569bf9](https://github.com/git/git/commit/5569bf9bbedd63a00780fc5c110e0cfab3aa97b9).
+{{% /caption %}}
+
+The third problem was related to how Hercules works: it is able to analyze any
+set of commits, not just the whole master. There is not guarantee that the commits
```
Comment (bzz) [inline content/post/hercules.v4.md]: `There is not guarantee` -> `There is no guarantee` ?

**Comment by vmarkovtsev** [comment]
Thanks @bzz staging updated

**Comment by vmarkovtsev** [comment]
@vcoisne @eiso Since Francesc went on vacations, maybe you will review this instead?

File: content/post/hercules.v4.md [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+Each pipeline item has a name and specifies its dependencies and what it provides in return.
+For example, a `RenameAnalysis` item detects the renamed and slightly changed files between two commits.
+It consumes the list of changes and the list of corresponding blobs. It calculates
+a better list of changes with some of the "delete + create" pairs replaced with
+single "edit under new name" elements. Yep, Git packfiles do not store this kind
+of information (intentionally) and `git` detects renames using heuristics every
+time you execute it.
+
+Hercules builds the DAG of pipeline items automatically based on their dependencies
+and provided intermediate results. Some items try to improve on the others -
```
Comment (campoy) [inline content/post/hercules.v4.md]: This sentence needs some work, it's too complex and hard to parse.
Also, what do you mean the items try to improve on the others? Which others?

Maybe:

> Each item on a pipeline transforms the result of its dependencies to generate its own result,
> as `RenameAnalysis` mentioned above.
> Hercules understands these dependencies and automatically sets up the corresponding computation graph.
> Here is an example of the built pipeline to calculate the burn down chart.

File: content/post/hercules.v4.md [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+pointed from the "improved" ones, and the "improved" ones in turn depend on the
+"basic" ones. Here is an example of the built pipeline to calculate the burndown
+stats:
+
+{{% caption src="/post/hercules.v4/pipeline.png" %}}
+Pipeline DAG example.
+{{% /caption %}}
+
+Once the pipeline is built, each item changes its state in certain ways.
+Here is the most full example of the "leaf" item which does not provide any
```
Comment (campoy) [inline content/post/hercules.v4.md]: I'm not sure I understand this sentence. What are those "certain ways"? Either drop it or explain it.
Also in what way this is the most full example? Maybe replace "the most full example" by "a complete example" or "an end-to-end example".

File: content/post/hercules.v4.md [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+intermediate results but rather a final result:
+
+{{% caption src="/post/hercules.v4/item.png" width="half-width" %}}
+Item lifecycle.
+{{% /caption %}}
+
+Each item is configured - its parameters are adjusted as needed. Returning
```
Comment (campoy) [inline content/post/hercules.v4.md]: Is this an explanation of the graph above? I don't see the relationship

File: content/post/hercules.v4.md [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+This versatile architecture allows Hercules to efficiently analyze small and big
+Git repositories while minimizing the complexity and boilerplate code to add
+new analysis types.
+
+# v4 challenges
+
+There were three most difficult technical challenges which had to be solved in v3 and v4:
+
+1. Merging results together.
+2. Forks and merges behavior of `PipelineItem`-s.
```
Comment (campoy) [inline content/post/hercules.v4.md]: drop the `-`, it renders well enough without it

File: content/post/hercules.v4.md [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+3. Determine the order in which to process commits, forks, merges and garbage collect branches.
```
Comment (campoy) [inline content/post/hercules.v4.md]: "garbage collects" or "performs garbage collection on"

File: content/post/hercules.v4.md [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+[One of the scariest functions I have ever written.](https://github.com/src-d/hercules/blob/99abfb229a880d55462eecd2373433df686f8fcc/leaves/burndown.go#L441).
+Regarding (2), it was needed to write the code to merge multiple line interval-marked
+files together and organize the work with a partially shared, partially copied state.
+Git merge does not necessarily contain two branches: it can be three, four, etc.
+
+{{% caption src="/post/hercules.v4/git_abstract.png" %}}
+Abstract from [git/git](https://github.com/git/git) history.
+{{% /caption %}}
+
+I did not come up with a solution better than mark the changed line intervals in files
```
Comment (campoy) [inline content/post/hercules.v4.md]: a better solution better than marking

File: content/post/hercules.v4.md [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+while consuming a merge commit with a special outlier value, then
+annotate each line of the each merged file, and copy the only real last edit date
+over those marked as outliers. In places of same edits but conflicting dates we
+choose the oldest date. In places of merge conflicts we automatically get the date
```
Comment (campoy) [inline content/post/hercules.v4.md]: We avoid conflicting edits by choosing the one with the latest date.

File: content/post/hercules.v4.md [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+of the merge commit without special tricks. We should not compare the files
```
Comment (campoy) [inline content/post/hercules.v4.md]: Not sure about what you mean with the merge conflicts here. Please rephrase.

File: content/post/hercules.v4.md [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+which are not involved in the merge, so we keep track of their hashes, we maintain
+them. I also had to rewrite the statistics storage logic which used to be too
```
Comment (campoy) [inline content/post/hercules.v4.md]: Replace all of the "I" with "we", this discourages others from contributing.

File: content/post/hercules.v4.md [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+related to minimizing the number of forks and merges, in particular removing
+the no-op fast-forward-like back edges of the DAG.
+
+{{% caption src="/post/hercules.v4/ff.gif" %}}
+Example of the removal of a fast-forward DAG edge.
+{{% /caption %}}
+
+We [topologically sort](https://en.wikipedia.org/wiki/Topological_sorting)
+the nodes of our DAG and traverse it starting from the root.
+Each time we suspect that an edge can be removed we traverse backward from the children
```
Comment (campoy) [inline content/post/hercules.v4.md]: backwards

File: content/post/hercules.v4.md [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+until we reach the node we have already visited. If that node is the same as the parent
+then indeed we can remove the edge.
+
+The second problem was that the Git history may have more than one root commit so
+we need to find the main root and take into account the side effects.
+Don't believe me? Here is how [gitk](https://git-scm.com/docs/gitk) was merged inside git/git:
```
Comment (campoy) [inline content/post/hercules.v4.md]: Wow, that's cool!
We should use this fact that "git trees can have multiple roots" in the promotion of this post, @vcoisne

File: content/post/hercules.v4.md [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+
+The fourth problem was to schedule the garbage collection - remove the
+clones which are no longer needed. We do a DAG traversal and mark the places
+where branches were last used, then we insert branch disposals after the resulting
+marks.
+
+It was very helpful to debug with [`git bisect`](https://git-scm.com/docs/git-bisect).
+Normally people bisect to find regressions; I bisected to find offending commits which broke Hercules.
+
+# You can help!
```
Comment (campoy) [inline content/post/hercules.v4.md]: We want your help!

File: content/post/hercules.v4.md [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+
+Hercules can do so many things better! Here is a list of cool features you can work on:
+
+* Count lines with [enry](https://github.com/src-d/enry) and plot charts by language.
```
Comment (campoy) [inline content/post/hercules.v4.md]: Counting

File: content/post/hercules.v4.md [hunk 25]
```diff
@@ -0,0 +1,295 @@
...
+* Apply PageRank to the structural edits and find the most impactful developers (aka [Sourcecred](https://github.com/sourcecred/sourcecred)).
```
Comment (campoy) [inline content/post/hercules.v4.md]: Applying ... (same below)

File: content/post/hercules.v4.md [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+* Detect minor commits using machine learning.
```
Comment (campoy) [inline content/post/hercules.v4.md]: What is a minor commit?

File: content/post/hercules.v4.md [hunk 27]
```diff
@@ -0,0 +1,295 @@
...
+* Plot with Go, not with Python.
+* Integrate into [GitBase](https://github.com/src-d/gitbase) as a set of User Defined Functions.
+* Do something with the awful plot design. [My colleague Maxim has attempted already.](https://github.com/smacker/hercules-web)
```
Comment (campoy) [inline content/post/hercules.v4.md]: Hahaha, make this a bit more positive?

> Improving the look of the generated plots.

File: content/post/hercules.v4.md (hunk 21 above)
Comment (vmarkovtsev) [reply content/post/hercules.v4.md]: @bzz haha!

Comment (vmarkovtsev) [reply content/post/hercules.v4.md]: https://github.com/src-d/blog/pull/239#discussion_r213355255

File: content/post/hercules.v4.md (hunk 12 above)
Comment (vmarkovtsev) [reply content/post/hercules.v4.md]: This simplifies the things but hell yeah better to simplify than to make more complex

File: content/post/hercules.v4.md (hunk 19 above)
Comment (vmarkovtsev) [reply content/post/hercules.v4.md]: Removed - I did not get me from the past too

**Comment by vmarkovtsev** [comment]
Thank you so much @campoy - all fixed. Staging updated.

**Comment by vcoisne** [comment]
LGTM

**Comment by vmarkovtsev** [comment]
@vcoisne in DM: LGTM
